storage/connectors/bigquery/src/main/java/feast/storage/connectors/bigquery/writer/BigQueryWrite.java

```diff
@@ -16,92 +16,227 @@
 import feast.storage.api.writer.WriteResult;
-import java.io.IOException;
-import org.apache.beam.sdk.io.gcp.bigquery.BigQueryIO;
-import org.apache.beam.sdk.io.gcp.bigquery.BigQueryInsertError;
-import org.apache.beam.sdk.io.gcp.bigquery.InsertRetryPolicy;
-import org.apache.beam.sdk.transforms.DoFn;
-import org.apache.beam.sdk.transforms.PTransform;
-import org.apache.beam.sdk.transforms.ParDo;
-import org.apache.beam.sdk.values.PCollection;
+import java.util.List;
+import java.util.Map;
+import org.apache.beam.sdk.coders.KvCoder;
+import org.apache.beam.sdk.coders.StringUtf8Coder;
+import org.apache.beam.sdk.io.gcp.bigquery.*;
+import org.apache.beam.sdk.transforms.*;
+import org.apache.beam.sdk.transforms.join.CoGbkResult;
+import org.apache.beam.sdk.transforms.join.CoGroupByKey;
+import org.apache.beam.sdk.transforms.join.KeyedPCollectionTuple;
+import org.apache.beam.sdk.transforms.windowing.FixedWindows;
+import org.apache.beam.sdk.transforms.windowing.Window;
+import org.apache.beam.sdk.values.*;
+import org.joda.time.Duration;
 import org.slf4j.Logger;
 
 /**
- * A {@link PTransform} that writes {@link FeatureRowProto FeatureRows} to the specified BigQuery
+ * A {@link PTransform} that writes {@link FeatureRow FeatureRows} to the specified BigQuery
  * dataset, and returns a {@link WriteResult} containing the unsuccessful writes. Since Bigquery
- * does not output successful writes, we cannot emit those, and so no success metrics will be
- * captured if this sink is used.
+ * does not output failed writes, we cannot emit those.
  */
-public class BigQueryWrite
-    extends PTransform<PCollection<FeatureRowProto.FeatureRow>, WriteResult> {
+public class BigQueryWrite extends PTransform<PCollection<FeatureRow>, WriteResult> {
   private static final Logger log = org.slf4j.LoggerFactory.getLogger(BigQueryWrite.class);
 
-  // Destination dataset
+  private static final Duration BIGQUERY_DEFAULT_WRITE_TRIGGERING_FREQUENCY =
+      Duration.standardMinutes(1);
+
+  private static final Duration BIGQUERY_JOB_MAX_EXPECTING_RESULT_TIME = Duration.standardHours(1);
+  private static final int BIGQUERY_MAX_JOB_RETRIES = 20;
+
   private DatasetId destination;
+  private PCollectionView<Map<String, Iterable<TableSchema>>> schemas;
 
-  public BigQueryWrite(DatasetId destination) {
+  private Duration triggeringFrequency = BIGQUERY_DEFAULT_WRITE_TRIGGERING_FREQUENCY;
+  private Duration expectingResultTime = BIGQUERY_JOB_MAX_EXPECTING_RESULT_TIME;
+  private BigQueryServices testServices;
+
+  public BigQueryWrite(
+      DatasetId destination, PCollectionView<Map<String, Iterable<TableSchema>>> schemas) {
     this.destination = destination;
+    this.schemas = schemas;
+  }
+
+  public BigQueryWrite withTriggeringFrequency(Duration triggeringFrequency) {
+    this.triggeringFrequency = triggeringFrequency;
+    return this;
+  }
+
+  public BigQueryWrite withTestServices(BigQueryServices testServices) {
```

Comment:
Should this be called something like `withBigQueryService` in case it is used outside the testing context?

Reply:
It's basically same method as here https://github.com/apache/beam/blob/b12a38aa2156c2652937669aeb2e46be15eca06c/sdks/java/io/google-cloud-platform/src/main/java/org/apache/beam/sdk/io/gcp/bigquery/BigQueryIO.java#L2388
and there's no other (public) implementation except the test one, and it was done intentionally